### PR TITLE
MIOpenTensile Part1 : Add MIOpenTensile in GEMM path

### PR DESCRIPTION
**Pull request by ce1adon**
The purpose of this PR is to add MIOpenTensile path in GEMM algorithm for fp32 on hip backend. It replaces the default GEMM path through rocblas in HIP backend for fp32 type.

### CODE REVIEW

**Comment by atamazov** [comment]
The Invokers work is complicated by frequent changes in the baseline code. For example, recent additions of ImplicitGemm solvers have created substantial additional work for @DrizztDoUrden, and, thus have delayed the overall Invoker's ETA.

@DrizztDoUrden is also reworking Gemm code to Solvers and Invokers. This work is currently paused because Vasya is doing something more important (ImplicitGemm Invokers), but I heard it is almost ready. If this PR will be merged first, then its code will need to be reworked once again almost immediately after merging, and that is we all would like to avoid, I guess.

I am not aware of the priorities of this MIOpenTensile thing. @daniellowell @ce1adon @DrizztDoUrden could you please communicate and decide the optimal order?

**Comment by DrizztDoUrden** [comment]
Doesn't matter much unless it's conv...ocl.cpp changes, which are not here. I don't change gemm internals there.

**Comment by atamazov** [comment]
Okay, then. @ce1adon Could you please set priority label.

**Comment by atamazov** [comment]
@ce1adon Please describe how the PR affects the library's behavior. Or this is NFC?

**Comment by ce1adon** [comment]
> @ce1adon Please describe how the PR affects the library's behavior. Or this is NFC?

It's not a new feature. The influence is updated in top comment.

**Comment by ce1adon** [comment]
Jira ticker related to the blocker
http://ontrack-internal.amd.com/browse/SWDEV-233998

File: src/include/miopen/solver.hpp
Comment (atamazov) [inline src/include/miopen/solver.hpp:1709]: Please revert this.

File: src/mlo_dir_conv.cpp
Comment (atamazov) [inline src/mlo_dir_conv.cpp:191]: Ditto

File: src/ocl/convolutionocl.cpp
Comment (atamazov) [inline src/ocl/convolutionocl.cpp:633]: What if miopentensile is not enabled?

File: src/solver.cpp
Comment (atamazov) [inline src/solver.cpp:234]: Please revert this

File: src/solver/conv_ocl_dir2D_bwdWrW_2.cpp
Comment (atamazov) [inline src/solver/conv_ocl_dir2D_bwdWrW_2.cpp:29]: Ditto

File: CMakeLists.txt
Comment (atamazov) [inline CMakeLists.txt:163]: Maybe
```suggestion
    # MIOpenTensile does not support opencl backend yet.
```

**Review by atamazov — CHANGES_REQUESTED** [review]
This change is NFC provided that `-DMIOPEN_USE_MIOPENTENSILE=On` is not used, right?

**Comment by atamazov** [comment]
ACHTUNG :warning: It seems like the behavior of GEMM algorithm, such as speed and applicability, is subject to change when MIOpenTensile is enabled. This affects the correctness of `find-db`. In turn, this may break the Immediate mode and all accelerated Find modes (including Hybrid which is the default now).

___We need to analyze the impact carefully.___ Until that, enabling MIOpenTensile can be used for experimental purposes only.

File: src/gemm_v2.cpp
Comment (ce1adon) [inline src/gemm_v2.cpp:204]: https://github.com/ROCmSoftwarePlatform/MIOpen/pull/139#discussion_r478724264

It will fall back in rocblas.

File: src/ocl/convolutionocl.cpp
Comment (ce1adon) [reply src/ocl/convolutionocl.cpp:633]: https://github.com/ROCmSoftwarePlatform/MIOpen/pull/139/files#r479479017

File: CMakeLists.txt
Comment (ce1adon) [reply CMakeLists.txt:163]: Will do

File: src/ocl/convolutionocl.cpp
Comment (atamazov) [reply src/ocl/convolutionocl.cpp:633]: I see. If so, then `GemmBackend_t::miopentensile` looks confusing. Shali we rename it to `GemmBackend_t::miopentensile_or_rocblas` and provide a comment saying that when/if miopentensile is not available, GEMM uses rocblas?

Comment (ce1adon) [reply src/ocl/convolutionocl.cpp:633]: Maybe not now. In fact we use the same rule for rocblas and miopengemm.

Comment (atamazov) [reply src/ocl/convolutionocl.cpp:633]: Okay. Please do that when you have time/will.

File: CMakeLists.txt
Comment (atamazov) [inline CMakeLists.txt:292]: I think that cmake should issue an error message and fail here. /cc  @pfultz2

**Review by atamazov — CHANGES_REQUESTED** [review]

**Comment by ce1adon** [comment]
Please review this PR.
I will revert the wrw2 changes and add a "tensile=OFF rocblas=ON" test when addressing the reviews.

File: CMakeLists.txt
Comment (ce1adon) [reply CMakeLists.txt:301]: Not really.
The failure was causing by dependency installation.

File: src/solver/conv_ocl_dir2D_bwdWrW_2.cpp
Comment (ce1adon) [reply src/solver/conv_ocl_dir2D_bwdWrW_2.cpp:29]: done

File: src/solver.cpp
Comment (ce1adon) [reply src/solver.cpp:234]: done

File: src/mlo_dir_conv.cpp
Comment (ce1adon) [reply src/mlo_dir_conv.cpp:191]: done

File: src/include/miopen/solver.hpp
Comment (ce1adon) [reply src/include/miopen/solver.hpp:1712]: done

File: CMakeLists.txt
Comment (atamazov) [reply CMakeLists.txt:297]: I mean that if the user wants to use miopentensile (`-DMIOPEN_USE_MIOPENTENSILE=On`) and package not found, then cmake should fail.

File: doc/src/DebugAndLogging.md
Comment (atamazov) [inline doc/src/DebugAndLogging.md:140]: Please use value 4. When env var is unset, `Value(ENV_VAR)` returns 0.

File: test/CMakeLists.txt
Comment (atamazov) [inline test/CMakeLists.txt:40]: [Question] Shall we automatically set it to "On" when `MIOPEN_USE_MIOPENTENSILE=On`?

**Review by atamazov — CHANGES_REQUESTED** [review]

File: test/CMakeLists.txt
Comment (ce1adon) [reply test/CMakeLists.txt:40]: I think we want them OFF as default. This mode is dedicated for GEMM related tests with miopentensile.

File: doc/src/DebugAndLogging.md
Comment (ce1adon) [reply doc/src/DebugAndLogging.md:140]: Will do

File: test/CMakeLists.txt
Comment (atamazov) [reply test/CMakeLists.txt:40]: Then in which cases and how we will enable these tests?

Comment (ce1adon) [reply test/CMakeLists.txt:40]: Right, looks like tests are disabled. Enabled in new commit.

File: Jenkinsfile
Comment (atamazov) [inline Jenkinsfile:580]: `-DMIOPEN_TEST_MIOTENSILE=ON` occurs twice in a line.

File: Jenkinsfile
Comment (atamazov) [inline Jenkinsfile:597]: Ditto

File: doc/src/DebugAndLogging.md
Comment (atamazov) [inline doc/src/DebugAndLogging.md:140]: Please maintain backward compatibility for the sake of who is already using this.

File: src/gemm_v2.cpp
Comment (atamazov) [inline src/gemm_v2.cpp:188]: Ditto

File: test/CMakeLists.txt
Comment (atamazov) [inline test/CMakeLists.txt:86]: Please add a comment (to the CmakeLists file) that explains why this new condition is needed.

**Review by atamazov — CHANGES_REQUESTED** [review]

File: Jenkinsfile
Comment (atamazov) [reply Jenkinsfile:580]: Oops, I am wrong.

File: doc/src/DebugAndLogging.md
Comment (ce1adon) [reply doc/src/DebugAndLogging.md:140]: done

File: src/gemm_v2.cpp
Comment (ce1adon) [reply src/gemm_v2.cpp:188]: done

File: test/CMakeLists.txt
Comment (ce1adon) [reply test/CMakeLists.txt:86]: It's not new. Please check the original file.
If this condition is missing, all the tests will be skipped because they are not in skip_all_expect_tests list. This condition restrict the usage of this list in certain tests.

File: CMakeLists.txt
Comment (ce1adon) [reply CMakeLists.txt:297]: Do you mean cmake should fail with error msg if MIOPEN_USE_MIOPENTENSILE is ON but package is not found.

Comment (ce1adon) [reply CMakeLists.txt:297]: Considering that the tensile eco is still under construction, I suggest not issuing error msg but just msg to warn user that it is not built with miopentensile. When miopentensile becomes the only gemm backend, we will issue the error msg. @atamazov @pfultz2

Comment (atamazov) [reply CMakeLists.txt:297]: I still think that when cmake is explicitly asked to do something but can't do that it should fail. I found this useful in real life. Please look at [similar case at initial COMGR PR](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/177#discussion_r416088277).

@pfultz2 What is your opinion?

File: test/CMakeLists.txt
Comment (atamazov) [reply test/CMakeLists.txt:86]: [don't resolve/hide conversation] Okay, resolved. But this stuff should be eventually rewritten. It is totally puzzling :exploding_head:

Comment (ce1adon) [reply test/CMakeLists.txt:86]: I can add a comment here.

Comment (atamazov) [reply test/CMakeLists.txt:86]: I would highly appreciate that. I still can't figure out how it works;)

Comment (ce1adon) [reply test/CMakeLists.txt:86]: done

**Comment by ce1adon** [comment]
OclBwdWrW53 issue
```
/var/jenkins/workspace/MLLibs_MIOpen_miotensile/build/bin/test_conv2d --float --cmode conv --pmode valid --group-count 1 --input 1, 16, 2048, 2048 --weights 1, 16, 5, 5 --pads_strides_dilations 0 0 1 1 1 1 --trans_output_pads 0 0 

FAILED: 0.00603448

Iteration: 61930

Backward weights convolution: ConvOclBwdWrW53

Input tensor: 1, 16, 2048, 2048

Weights tensor: 1, 16, 5, 5

Output tensor: 1, 1, 2044, 2044

Filter: conv2d, miopenConvolution, miopenPaddingValid, {0, 0}, {1, 1}, {1, 1}, 

Max diff: 5642

Mismatch at 0: 21546 != 23143
```

**Comment by ce1adon** [comment]
@atamazov @DrizztDoUrden @pfultz2 @daniellowell  @asroy 
This PR is ready to review.

File: Jenkinsfile
Comment (atamazov) [inline Jenkinsfile:303]: Ditto

File: CMakeLists.txt
Comment (atamazov) [reply CMakeLists.txt:299]: Okay, let's keep it for now. However please keep in mind that in case of missing miopentensile in out CI environment, the tensile tests would pass, but actual testing wouldn't happen.

File: src/solver/conv_ocl_dir2D_bwdWrW_2.cpp
Comment (atamazov) [inline src/solver/conv_ocl_dir2D_bwdWrW_2.cpp:481]: [Notice] IIRC this workaround should be removed prior merging.

File: src/solver/conv_multipass_wino3x3WrW.cpp
Comment (atamazov) [inline src/solver/conv_multipass_wino3x3WrW.cpp:367]: [minor issue] IIRC miopentensile does not support FP16 and BF16. That means that if `MIOPEN_USE_ROCBLAS == false` and `MIOPEN_USE_MIOPENTENSILE == true`, then this IsApplicable() should return `false` for FP16 and BF16.

**Review by atamazov — APPROVED** [review]
Minor issue, but I am not against merging this provided that MIOpenTenslie integration work will be continued soon.

File: src/solver/conv_ocl_dir2D_bwdWrW_2.cpp
Comment (ce1adon) [reply src/solver/conv_ocl_dir2D_bwdWrW_2.cpp:481]: Is it fixed?

Comment (atamazov) [reply src/solver/conv_ocl_dir2D_bwdWrW_2.cpp:481]: No. There is workaround PR #400 that seems stuck in Draft state. Let's discuss in https://github.com/AMDComputeLibraries/MLOpen/issues/2534.

Comment (atamazov) [reply src/solver/conv_ocl_dir2D_bwdWrW_2.cpp:481]: Please let me know if you need assistance from me.

Comment (ce1adon) [reply src/solver/conv_ocl_dir2D_bwdWrW_2.cpp:481]: I will check in some new changes, meanwhile I will remove this workaround. If the issue is still there, I will request your help.

**Comment by atamazov** [comment]
Putting on hold. See https://github.com/AMDComputeLibraries/MLOpen/issues/2534#issuecomment-699075454.

**Comment by ce1adon** [comment]
> Putting on hold. See [AMDComputeLibraries/MLOpen#2534 (comment)](https://github.com/AMDComputeLibraries/MLOpen/issues/2534#issuecomment-699075454).

We should actively test this PR at this moment, no matter if this bug exists. I will remove the on hold.

**Comment by atamazov** [comment]
@ce1adon Up to you. However, AFAIU `on_hold` does not prevent anything. It simply says that PR should not be merged.

File: src/solver/conv_ocl_dir2D_bwdWrW_2.cpp
Comment (ce1adon) [reply src/solver/conv_ocl_dir2D_bwdWrW_2.cpp:481]: done

File: src/solver/conv_multipass_wino3x3WrW.cpp
Comment (ce1adon) [reply src/solver/conv_multipass_wino3x3WrW.cpp:367]: done

File: src/solver/conv_multipass_wino3x3WrW.cpp
Comment (atamazov) [inline src/solver/conv_multipass_wino3x3WrW.cpp:367]: This (and next change) looks like unnecessary complication. Simply insert this:
```C++
/// miopentensile currently supports only FP32.
/// \todo Remove this when FP16/BF16 support on miopentensile path is ready.
#if !MIOPEN_USE_ROCBLAS
    if(!params.IsFp32())
        return false;
```

File: src/solver/conv_multipass_wino3x3WrW.cpp
Comment (atamazov) [inline src/solver/conv_multipass_wino3x3WrW.cpp:550]: This is not needed, please revert.

File: src/solver/conv_multipass_wino3x3WrW.cpp
Comment (atamazov) [inline src/solver/conv_multipass_wino3x3WrW.cpp:749]: Ditto.

**Review by atamazov — CHANGES_REQUESTED** [review]

File: src/solver/conv_multipass_wino3x3WrW.cpp
Comment (ce1adon) [reply src/solver/conv_multipass_wino3x3WrW.cpp:367]: will do

File: src/solver/conv_multipass_wino3x3WrW.cpp
Comment (ce1adon) [reply src/solver/conv_multipass_wino3x3WrW.cpp:550]: Do you mean remove both #if? The first condition was originally there for rocblas. Was the related workaround been removed?

File: src/solver/conv_multipass_wino3x3WrW.cpp
Comment (ce1adon) [reply src/solver/conv_multipass_wino3x3WrW.cpp:749]: will do

File: hip-clang.docker
Comment (pfultz2) [inline hip-clang.docker:71]: `-DAMDGPU_TARGETS=${GPU_ARCH}` should be passed `cget init`. Also this should install tensile from a file so that hcc and hip-clang install the same version.

Comment (ce1adon) [reply hip-clang.docker:71]: Will move gpu_arch to init.
Currently both hcc and hip-clang are using the same version. The version divergence here is both for hip-clang. It intends to test both a stable version and a latest version of yaml+tensile.

Comment (ce1adon) [reply hip-clang.docker:71]: moved gpu_arch flag to cget init

File: src/solver/conv_multipass_wino3x3WrW.cpp
Comment (ce1adon) [reply src/solver/conv_multipass_wino3x3WrW.cpp:546]: Revised, not reverted yet

**Comment by ce1adon** [comment]
The last test passed but is marked as failed.

**Comment by ce1adon** [comment]
NOTE: https://github.com/ROCmSoftwarePlatform/MIOpen/pull/139/commits/d0969a05ddd5533fcf81e6acdbd54e49674b1298 passed at # 38

**Review by atamazov — APPROVED** [review]
LGTM

File: test/CMakeLists.txt
Comment (atamazov) [inline test/CMakeLists.txt:221]: @ce1adon It should be
```
OR (MIOPEN_TEST_MIOTENSILE AND NOT (NAME IN_LIST SKIP_TESTS)))
```
Current code enables ALL tests when MIOPEN_TEST_MIOTENSILE=off, thus affecting non-miopentensile tests (half etc). I will fix that.

Comment (ce1adon) [reply test/CMakeLists.txt:221]: This is under the condition of (NOT (MIOPEN_TEST_INT8 OR MIOPEN_TEST_BFLOAT16 OR MIOPEN_TEST_HALF OR MIOPEN_TEST_MIOTENSILE)), which means this only applies to fp32 cases. Do you mean fp32 doesn't need to test all the cases?

Comment (atamazov) [reply test/CMakeLists.txt:221]: No, this is not "under" that confition. It is just one of ORed conditions. Alternate formatting:
```
if(
    (NOT (MIOPEN_TEST_INT8 OR MIOPEN_TEST_BFLOAT16 OR MIOPEN_TEST_HALF OR MIOPEN_TEST_MIOTENSILE)) // (1)
    OR (MIOPEN_TEST_BFLOAT16 AND ${PARSE_ALLOW_BFLOAT16}) // (2)
    OR ( MIOPEN_TEST_HALF AND ${PARSE_ALLOW_HALF}) // (3)
    OR (MIOPEN_TEST_INT8 AND ${PARSE_ALLOW_INT8}) // (4)
    OR (NOT (MIOPEN_TEST_MIOTENSILE AND (NAME IN_LIST SKIP_TESTS))) // (5)
) ...
```
When MIOPEN_TEST_MIOTENSILE=Off, condition (5) is always true. This works properly for MIOpenTensile, that's why you've missed that, I guess.

Comment (atamazov) [reply test/CMakeLists.txt:221]: Fixed in 318f48073e1bee6d1 (not yet merged to develop)

Comment (atamazov) [reply test/CMakeLists.txt:221]: Fixed in #750

**Comment by atamazov** [comment]
@ce1adon 
> It replaces the default GEMM path through rocblas in HIP backend for fp32 type.

If MIOpenTensile is the default for FP32, then shall we remove FP32 testing from "MIOpenTensile" stages in the Jenkinsfile?

**Pull request merged**
